File: backend/layers/features/workbench/workbench_service.py
from __future__ import annotations

from typing import Any, Protocol

from backend.layers.common.governance.lifecycle import parse_positive_integer


class WorkbenchServiceError(ValueError):
    def __init__(self, code: str, message: str, status: int = 400) -> None:
        super().__init__(message)
        self.code = code
        self.status = status


class WorkbenchStore(Protocol):
    def workbench_summary(self, *, user: dict[str, Any]) -> dict[str, Any]: ...
    def list_work_items(self, *, user: dict[str, Any], status: str | None, include_history: bool, page: int, page_size: int) -> dict[str, Any]: ...
    def transition_work_item(self, item_id: int, *, user: dict[str, Any], action: str, expected_version: int | None, note: str | None) -> dict[str, Any]: ...
    def list_notifications(self, *, user_id: int, status: str | None, include_history: bool, page: int, page_size: int) -> dict[str, Any]: ...
    def mark_notification_read(self, notification_id: int, *, user_id: int) -> dict[str, Any]: ...
    def close_notification(self, notification_id: int, *, user_id: int, conclusion: str) -> dict[str, Any]: ...


class WorkbenchService:
    def __init__(self, store: WorkbenchStore) -> None:
        self.store = store

    @staticmethod
    def require_permission(user: dict[str, Any], permission: str) -> None:
        permissions = set(user.get("permissions") or [])
        if permission not in permissions:
            raise WorkbenchServiceError("FORBIDDEN", "当前账号没有待办/消息权限", 403)

    def list_work_items(self, user: dict[str, Any], *, status: str | None = None, include_history: bool = True, page: int = 1, page_size: int = 20) -> dict[str, Any]:
        self.require_permission(user, "work_item.view")
        result = self.store.list_work_items(user=user, status=status, include_history=include_history, page=max(1, page), page_size=min(100, max(1, page_size)))
        return {**result, "items": [{**item, "handling_mode": "manual" if item.get("module_code") == "workbench" and item.get("object_type") == "workbench:manual" else "domain"} for item in result["items"]]}

    def summary(self, user: dict[str, Any]) -> dict[str, Any]:
        self.require_permission(user, "workbench.enter")
        summary = self.store.workbench_summary(user=user)
        work_items = self.store.list_work_items(user=user, status=None, include_history=False, page=1, page_size=5)
        notifications = self.store.list_notifications(user_id=int(user["id"]), status=None, include_history=False, page=1, page_size=5)
        summary["kpis"]["todo_open"] = int(work_items["total"])
        summary["todos"] = [
            {
                "id": item["id"], "title": item["title"], "type": item["action_code"],
                "due_at": item.get("due_at") or "未设置期限", "overdue": bool(item.get("overdue")),
            }
            for item in work_items["items"]
        ]
        summary["alerts"] = [
            {
                "id": item["id"], "title": item["title"],
                "level": "high" if item["level"] in {"high", "critical"} else "medium",
                "created_at": item.get("last_occurred_at") or item.get("created_at"),
            }
            for item in notifications["items"]
        ]
        return summary

    def transition_work_item(self, user: dict[str, Any], item_id: int, *, action: str, expected_version: Any = None, note: str | None = None) -> dict[str, Any]:
        self.require_permission(user, "work_item.manage")
        if action not in {"claim", "start", "complete", "cancel"}:
            raise WorkbenchServiceError("VALIDATION_ERROR", "待办操作无效", 400)
        if action == "cancel" and not str(note or "").strip():
            raise WorkbenchServiceError("CANCEL_REASON_REQUIRED", "取消待办必须填写原因", 400)
        normalized_version = None if expected_version is None else parse_positive_integer(
            expected_version,
            code="EXPECTED_VERSION_REQUIRED",
            message="必须提供 expected_version",
        )
        try:
            return self.store.transition_work_item(int(item_id), user=user, action=action, expected_version=normalized_version, note=str(note).strip() if note else None)
        except PermissionError as exc:
            raise WorkbenchServiceError("FORBIDDEN", str(exc), 403) from exc
        except ValueError as exc:
            raise WorkbenchServiceError("WORK_ITEM_TRANSITION_FAILED", str(exc), 409) from exc

    def list_notifications(self, user: dict[str, Any], *, status: str | None = None, include_history: bool = True, page: int = 1, page_size: int = 20) -> dict[str, Any]:
        self.require_permission(user, "work_item.view")
        return self.store.list_notifications(user_id=int(user["id"]), status=status, include_history=include_history, page=max(1, page), page_size=min(100, max(1, page_size)))

    def mark_notification_read(self, user: dict[str, Any], notification_id: int) -> dict[str, Any]:
        self.require_permission(user, "work_item.manage")
        try:
            return self.store.mark_notification_read(int(notification_id), user_id=int(user["id"]))
        except ValueError as exc:
            raise WorkbenchServiceError("NOTIFICATION_UPDATE_FAILED", str(exc), 409) from exc

    def close_notification(self, user: dict[str, Any], notification_id: int, conclusion: str) -> dict[str, Any]:
        self.require_permission(user, "work_item.manage")
        if not conclusion.strip():
            raise WorkbenchServiceError("CLOSE_CONCLUSION_REQUIRED", "关闭消息必须填写处理结论", 400)
        try:
            return self.store.close_notification(int(notification_id), user_id=int(user["id"]), conclusion=conclusion.strip())
        except ValueError as exc:
            raise WorkbenchServiceError("NOTIFICATION_UPDATE_FAILED", str(exc), 409) from exc

File: frontend/src/layers/features/workbench/workbench.service.ts
import { createApiClient } from '../../common/api/client'
import type { PageResult, PondDetail, PondSummary, WorkbenchSummary } from '../../common/api/workbench.models'
import type { MasterPage, MasterRecord } from '../../common/api/master-data.models'

const api = createApiClient()

export interface PondListQuery { search?: string; status?: string; area_id?: number; page?: number; page_size?: number }

export const getWorkbenchSummary = () => api.get<WorkbenchSummary>('/api/v1/workbench/summary')

function pond(row: MasterRecord): PondSummary {
  return {
    ...row,
    pond_code: row.code,
    area_name: String(row.area_name ?? `区域 ${row.area_id ?? '—'}`),
    area_id: Number(row.area_id ?? 0),
    group_name: String(row.group_name ?? '未分组'),
    group_id: Number(row.pond_group_id ?? 0),
    status: String(row.pond_status ?? 'build') as PondSummary['status'],
    lifecycle_status: row.status as PondSummary['lifecycle_status'],
    capacity_mu: Number(row.capacity_mu ?? 0),
    species: String(row.species ?? '待定'),
    active_batch_count: Number(row.active_batch_count ?? 0),
    updated_at: String(row.updated_at ?? ''),
  }
}

export async function listPonds(query: PondListQuery = {}): Promise<PageResult<PondSummary>> {
  const params = new URLSearchParams(Object.entries(query).filter(([, value]) => value !== undefined && value !== '').map(([key, value]) => [key, String(value)]))
  const result = await api.get<MasterPage>(`/api/v1/master-data/ponds?${params}`)
  return { ...result, items: result.items.map(pond) }
}

export async function getPond(id: number): Promise<PondDetail> {
  const result = await api.get<{ record: MasterRecord }>(`/api/v1/master-data/ponds/${id}`)
  const normalized = pond(result.record)
  return {
    ...normalized,
    manager_name: String(result.record.manager_name ?? '—'),
    water_source: String(result.record.water_source ?? '—'),
    location: String(result.record.location_text ?? '—'),
    notes: String(result.record.description ?? '—'),
    aerator_count: result.record.aerator_count == null ? undefined : Number(result.record.aerator_count),
    stocking_spec: result.record.stocking_spec == null ? undefined : String(result.record.stocking_spec),
    current_spec: result.record.current_spec == null ? undefined : String(result.record.current_spec),
    stock_quantity: result.record.stock_quantity == null ? undefined : Number(result.record.stock_quantity),
    stock_quantity_source: result.record.stock_quantity_source == null ? undefined : String(result.record.stock_quantity_source),
    timeline_preview: Array.isArray(result.record.timeline_preview) ? result.record.timeline_preview as PondDetail['timeline_preview'] : [],
    status_change_targets: Array.isArray(result.record.status_change_targets) ? result.record.status_change_targets as PondDetail['status_change_targets'] : [],
    can_request_status_change: result.record.can_request_status_change === true,
    can_verify_status_change: result.record.can_verify_status_change === true,
    pending_status_change: result.record.pending_status_change ? result.record.pending_status_change as unknown as PondDetail['pending_status_change'] : null,
  }
}

export interface WorkItemRecord {
  id: number
  title: string
  detail?: string | null
  module_code: string
  action_code: string
  object_type?: string | null
  object_id?: number | null
  object_ref?: string | null
  handling_mode?: 'manual' | 'domain'
  priority: 'low' | 'normal' | 'high' | 'critical'
  status: 'pending' | 'claimed' | 'in_progress' | 'completed' | 'cancelled' | 'escalated'
  due_at?: string | null
  overdue?: boolean
  completed_at?: string | null
  completion_note?: string | null
  cancelled_at?: string | null
  cancel_reason?: string | null
  row_version: number
}

export interface NotificationRecord {
  id: number
  title: string
  body?: string | null
  module_code: string
  level: 'low' | 'normal' | 'high' | 'critical'
  status: 'unread' | 'read' | 'closed' | 'escalated'
  last_occurred_at?: string | null
  read_at?: string | null
  closed_at?: string | null
  close_conclusion?: string | null
  occurrence_count: number
  object_type?: string | null
  object_id?: number | null
  object_ref?: string | null
}

export interface GovernanceList<T> { items: T[]; page: number; page_size: number; total: number; has_next: boolean }

export function getWorkItems(includeHistory = true, page = 1, pageSize = 100): Promise<GovernanceList<WorkItemRecord>> { return api.get(`/api/v1/work-items?include_history=${includeHistory ? 'true' : 'false'}&page=${page}&page_size=${pageSize}`) }
export function transitionWorkItem(id: number, action: 'claim' | 'start' | 'complete' | 'cancel', expectedVersion: number, note?: string): Promise<{ work_item: WorkItemRecord }> { return api.patch(`/api/v1/work-items/${id}`, { action, expected_version: expectedVersion, note }) }
export function getNotifications(includeHistory = true, page = 1, pageSize = 100): Promise<GovernanceList<NotificationRecord>> { return api.get(`/api/v1/notifications?include_history=${includeHistory ? 'true' : 'false'}&page=${page}&page_size=${pageSize}`) }
export function updateNotification(id: number, status: 'read' | 'closed', conclusion?: string): Promise<{ notification: NotificationRecord }> { return api.patch(`/api/v1/notifications/${id}`, { status, conclusion }) }
